src/v/storage/segment_utils.cc

```diff
@@ -87,12 +87,41 @@ static inline ss::file_open_options writer_opts() {
     };
 }
 
+/// Check if the file is on BTRFS, and disable copy-on-write if so.  COW
+/// is not useful for logs and can cause issues.
+static ss::future<>
+maybe_disable_cow(const std::filesystem::path& path, ss::file& file) {
+    if (co_await ss::file_system_at(path.string()) == ss::fs_type::btrfs) {
+        try {
+            int flags = -1;
+            // ss::syscall_result throws on errors, so not checking returns.
+            co_await file.ioctl(FS_IOC_GETFLAGS, (void*)&flags);
+            if ((flags & FS_NOCOW_FL) == 0) {
+                flags |= FS_NOCOW_FL;
+                co_await file.ioctl(FS_IOC_SETFLAGS, (void*)&flags);
+                vlog(stlog.trace, "Disabled COW on BTRFS segment {}", path);
+            }
+        } catch (std::system_error& e) {
+            // Non-fatal, user will just get degraded behaviour
+            // when btrfs tries to COW on a journal.
+            vlog(stlog.info, "Error disabling COW on {}: {}", path, e);
+        }
+    }
+}
+
 /// make file handle with default opts
 ss::future<ss::file> make_writer_handle(
   const std::filesystem::path& path, storage::debug_sanitize_files debug) {
-    return make_handle(
+    std::filesystem::path path_copy = path;
```

Comment:
This is weird, it is crashing if I don't make a copy of the path before passing it into `make_handle`, but I odn't see what inside `make_handle` could be destroying the original

Reply:
paths are cheap, why not simply remove the const-ref from the func. could be schedule in a way by the coro compiler state machine rewrite where the second frame is suspended before you have access to the var. making a copy makes sense to me. it has a small-array optimization anyway, so should be cheap